src/scales/scale.category.js

```diff
@@ -63,68 +63,58 @@ module.exports = Scale.extend({
 		return me.ticks[index - me.minIndex];
 	},
 
-	// Used to get data value locations.  Value can either be an index or a numerical value
-	getPixelForValue: function(value, index, datasetIndex) {
+	_getParams: function() {
 		var me = this;
 		var offset = me.options.offset;
 
 		// 1 is added because we need the length but we have the indexes
-		var offsetAmt = Math.max(me.maxIndex + 1 - me.minIndex - (offset ? 0 : 1), 1);
+		var range = Math.max(me.maxIndex + 1 - me.minIndex - (offset ? 0 : 1), 1);
+		var start = me.minIndex - (offset ? 0.5 : 0);
+		return {
+			start: start,
+			range: range
+		};
+	},
 
-		var isHorizontal = me.isHorizontal();
-		var valueDimension = (isHorizontal ? me.width : me.height) / offsetAmt;
-		var valueCategory, labels, idx, pixel;
+	// Used to get data value locations.  Value can either be an index or a numerical value
+	getPixelForValue: function(value, index, datasetIndex) {
+		var me = this;
+		var params = me._getParams();
+		var valueCategory, labels, idx;
 
 		if (!isNullOrUndef(index) && !isNullOrUndef(datasetIndex)) {
 			value = me.chart.data.datasets[datasetIndex].data[index];
 		}
 
 		// If value is a data object, then index is the index in the data array,
 		// not the index of the scale. We need to change that.
-		if (!isNullOrUndef(value)) {
-			valueCategory = isHorizontal ? value.x : value.y;
+		if (!helpers.isNullOrUndef(value)) {
+			valueCategory = me.isHorizontal() ? value.x : value.y;
 		}
 		if (valueCategory !== undefined || (value !== undefined && isNaN(index))) {
 			labels = me._getLabels();
 			value = helpers.valueOrDefault(valueCategory, value);
 			idx = labels.indexOf(value);
 			index = idx !== -1 ? idx : index;
+			if (isNaN(index)) {
+				index = value;
+			}
 		}
-
-		pixel = valueDimension * (index - me.minIndex);
-
-		if (offset) {
-			pixel += valueDimension / 2;
-		}
-
-		return (isHorizontal ? me.left : me.top) + pixel;
+		return me.getPixelForDecimal((index - params.start) / params.range);
 	},
 
 	getPixelForTick: function(index) {
-		return this.getPixelForValue(this.ticks[index], index + this.minIndex);
+		var ticks = this.ticks;
+		if (index < 0 || index > ticks.length - 1) {
+			return null;
+		}
+		return this.getPixelForValue(ticks[index], index + this.minIndex);
 	},
 
 	getValueForPixel: function(pixel) {
 		var me = this;
-		var offset = me.options.offset;
-		var offsetAmt = Math.max(me._ticks.length - (offset ? 0 : 1), 1);
-		var isHorizontal = me.isHorizontal();
-		var valueDimension = (isHorizontal ? me.width : me.height) / offsetAmt;
-		var value;
-
-		pixel -= isHorizontal ? me.left : me.top;
-
-		if (offset) {
-			pixel -= valueDimension / 2;
-		}
-
-		if (pixel <= 0) {
-			value = 0;
-		} else {
-			value = Math.round(pixel / valueDimension);
-		}
-
-		return value + me.minIndex;
+		var params = me._getParams();
+		return Math.round(params.start + me.getDecimalForPixel(pixel) * params.range);
```

Comment:
The current code is filtering out negative values. For safety, why don't we check the boundary by `Math.min(Math.max(value, 0), me._ticks.length - 1)`?